src/BuiltInTools/DotNetDeltaApplier/HotReloadAgent.cs

```diff
@@ -45,116 +44,211 @@ private void OnAssemblyLoad(object? _, AssemblyLoadEventArgs eventArgs)
+                    updateMethod.ReturnType == typeof(void))
+                {
+                    return updateMethod;
+                }
+
+                foreach (MethodInfo method in handlerType.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.Instance))
+                {
+                    if (method.Name == name)
                     {
-                        _log($"No BeforeUpdate or AfterUpdate method found on '{handlerType}'.");
+                        _log($"Type '{handlerType}' has method '{method}' that does not match the required signature.");
+                        break;
                     }
+                }
+
+                return null;
+            }
+        }
+
+        internal static List<Assembly> TopologicalSort(ReadOnlySpan<Assembly> assemblies)
+        {
+            var sortedAssemblies = new List<Assembly>(assemblies.Length);
 
-                    Action<Type[]?> CreateAction(MethodInfo update)
+            var visited = new HashSet<string>(StringComparer.Ordinal);
+
+            foreach (var assembly in assemblies)
+            {
+                Visit(assemblies, assembly, sortedAssemblies, visited);
+            }
+
+            static void Visit(ReadOnlySpan<Assembly> assemblies, Assembly assembly, List<Assembly> sortedAssemblies, HashSet<string> visited)
+            {
+                var assemblyIdentifier = assembly.FullName ?? assembly.ToString();
+                if (!visited.Add(assemblyIdentifier))
+                {
+                    return;
+                }
+
+                foreach (var dependencyName in assembly.GetReferencedAssemblies())
+                {
+                    var dependency = FindDependency(assemblies, dependencyName);
+                    if (dependency is not null)
                     {
-                        Action<Type[]?> action = update.CreateDelegate<Action<Type[]?>>();
-                        return types =>
-                        {
-                            try
-                            {
-                                action(types);
-                            }
-                            catch (Exception ex)
-                            {
-                                _log($"Exception from '{action}': {ex}");
-                            }
-                        };
+                        Visit(assemblies, dependency, sortedAssemblies, visited);
                     }
+                }
+
+                sortedAssemblies.Add(assembly);
+            }
 
-                    MethodInfo? GetUpdateMethod(Type handlerType, string name)
+            static Assembly? FindDependency(ReadOnlySpan<Assembly> assemblies, AssemblyName dependencyName)
+            {
+                foreach (var assembly in assemblies)
+                {
+                    if (assembly.FullName == dependencyName.FullName)
```

Comment:
The dependencies should be matched using simple name. FullName includes version that will often not match.

Reply:
I started with that, but I was worried because the nullability annotations suggested that the name could be null. Does it follow that if FullName is non-null for runtime loaded assemblies, `GetName().Name` is also non-null?

Reply:
Yes, `GetName().Name` is always going to be non-null for runtime loaded assemblies.

Reply:
Do you have suggestions on how we would approach a graph like this:

```
[
    "System.Private.CoreLib": /* depends on */ [] ,
    "System.Text.Json": ["System.Text.PrivateCoreLib"],
    "Microsoft.AspNetCore.Components, Version=3.1.0.0": ["System.Text.PrivateCoreLib"],
    "Microsoft.AspNetCore.Components, Version=6.0.0.0": ["System.Text.PrivateCoreLib", "System.Text.Json"],
    "Microsoft.AspNetCore.Components.Web, Version=6.0.0.0": ["Microsoft.AspNetCore.Components, Version=6.0.0.0"],
]
```

Relying on simple name might result in `Microsoft.AspNetCore.Components.Web` to be sorted alongside `System.Text.Json`.

Reply:
You should not see both `Microsoft.AspNetCore.Components, Version=6.0.0.0` and `Microsoft.AspNetCore.Components, Version=3.1.0.0` at runtime. The 3.1 version will be unified with the 6.0 version. The runtime will only load  `Microsoft.AspNetCore.Components, Version=6.0.0.0` and it will be used to satisfy both 3.1 and 6.0 references. 

This is assuming default build and loader behavior. If somebody has a custom solutions using AssemblyLoadContexts and multiple versions of the same thing loaded SxS, it gets really complex. I do not think we need to worry about cases like that for hot reload.